src/interchange/src/json.rs

```diff
@@ -158,14 +159,20 @@ impl ToJson for TypedDatum<'_> {
             // https://stackoverflow.com/questions/10286204/what-is-the-right-json-date-format
             ScalarType::Date => serde_json::Value::String(format!("{}", datum.unwrap_date())),
             ScalarType::Time => serde_json::Value::String(format!("{:?}", datum.unwrap_time())),
-            ScalarType::Timestamp { .. } => serde_json::Value::String(format!(
-                "{:?}",
-                datum.unwrap_timestamp().to_naive().timestamp_millis()
-            )),
-            ScalarType::TimestampTz { .. } => serde_json::Value::String(format!(
-                "{:?}",
-                datum.unwrap_timestamptz().to_naive().timestamp_millis()
-            )),
+            ScalarType::Timestamp { .. } => {
+                let naive = datum.unwrap_timestamp().to_naive();
+                let millis = naive.timestamp_millis();
+                let micros = naive.timestamp_subsec_micros()
+                    - (naive.timestamp_subsec_millis() * MICROS_PER_MILLIS);
+                serde_json::Value::String(format!("{millis}.{micros:0>3}"))
```

Comment:
I'm not in love with using string padding to get the decimal precision right, so if anyone has any better ideas let me know. As the commit message indicates, I tried using `Numeric` but it was 2-4x slower. I wanted to avoid using `f64` to avoid float precision issues.

Reply:
I think it's fine, but it should also be fine to just do `format!("{:.3}", (naive.timestamp_micros() as f64) / 1000.0)`.

This test program doesn't panic:

```
fn main() {
    for i in 0..1000 {
        let s1 = format!("{:.3}", (i as f64) / 1000.);
        let s2 = format!("0.{i:0>3}");
        assert_eq!(s1, s2);
    }
}
```

Reply:
Oh I see, since we know that `micros` will be between 0 and 1000, we know that this won't be lossy. Nice, I'll make the update.

Reply:
Oh wait, I don't think that's actually right. `naive.timestamp_micros()` will return the current epoch, which is currently something like `1709052306000000`. So there still might be a risk that `naive.timestamp_micros() as f64` will be a lossy conversion and silently change the value of the timestamp. Or in practice is it not an issue for `f64` with the realistic range of the current epoch?

Reply:
> Or in practice is it not an issue for f64 with the realistic range of the current epoch?

I _think_ so, but I don't know enough about floats to actually be 100% sure this is okay, so you're probably better off doing it in your original way to be safe.